File: TAC/Demo/DemoTac.java
import java.sql.*;
import java.time.Instant;
import java.util.Properties;
import oracle.ucp.jdbc.PoolDataSource;
import oracle.ucp.jdbc.PoolDataSourceFactory;

public class DemoTacUcp {
  public static void main(String[] args) throws Exception {
    String url =
      "jdbc:oracle:thin:@" +
      "(DESCRIPTION=(CONNECT_TIMEOUT=20)(TRANSPORT_CONNECT_TIMEOUT=3)(RETRY_COUNT=60)(RETRY_DELAY=2)" +
      " (ADDRESS_LIST=(LOAD_BALANCE=ON)" +
      "   (ADDRESS=(PROTOCOL=TCP)(HOST=<PRI_HOST>)(PORT=1521))" +
      "   (ADDRESS=(PROTOCOL=TCP)(HOST=<STBY_HOST>)(PORT=1521))" +
      " )(CONNECT_DATA=(SERVICE_NAME=br_tac_svc)))";

    PoolDataSource pds = PoolDataSourceFactory.getPoolDataSource();
    pds.setConnectionFactoryClassName("oracle.jdbc.pool.OracleDataSource");
    pds.setURL(url);
    pds.setUser("<USER>");
    pds.setPassword("<PASSWORD>");
    pds.setFastConnectionFailoverEnabled(true);   // FAN/FCF on
    pds.setValidateConnectionOnBorrow(true);
    // Optional sizings
    pds.setInitialPoolSize(1);
    pds.setMinPoolSize(1);
    pds.setMaxPoolSize(5);

    String sql =
      "MERGE INTO demo_tac_ac t " +
      "USING (SELECT ? id, ? note FROM dual) s " +
      "ON (t.id = s.id) " +
      "WHEN NOT MATCHED THEN INSERT (id, note) VALUES (s.id, s.note)";

    int id = 1001;
    System.out.println("Running… update every 1s. Perform a switchover now.");

    while (true) {
      String note = "JAVA-TAC " + Instant.now().toString();
      try (Connection conn = pds.getConnection();
           PreparedStatement ps = conn.prepareStatement(sql)) {
        conn.setAutoCommit(false);
        ps.setInt(1, id);
        ps.setString(2, note);
        ps.executeUpdate();
        conn.commit();
        System.out.println("Upserted id=" + id + " note=" + note);
      }
      Thread.sleep(1000);
    }
  }
}

File: TAC/Demo/demo_tac.py
import time, datetime, oracledb

# Initialize Thick mode once in your app startup (if not already done)
# oracledb.init_oracle_client(lib_dir="/opt/oracle/instantclient_19_20")

conn = oracledb.connect(user="<USER>", password="<PASSWORD>",
                        dsn="//<SCAN-or-host>:1521/br_tac_svc")  # service required
conn.autocommit = False

sql = ( "MERGE INTO demo_tac_ac t "
        "USING (SELECT :id id, :note note FROM dual) s "
        "ON (t.id = s.id) "
        "WHEN NOT MATCHED THEN INSERT (id, note) VALUES (s.id, s.note)" )

cur = conn.cursor()
id_val = 1003

print("Running… update every 1s. Perform a switchover now.")
while True:
    note = f"PY-TAC {datetime.datetime.utcnow().isoformat()}Z"
    cur.execute(sql, [id_val, note])  # replay-safe request
    conn.commit()                      # commit outcome enforced by service
    print(f"Upserted id={id_val} note={note}")
    time.sleep(1)
